src/arch/mod.rs

```diff
@@ -0,0 +1,58 @@
+#![allow(unsafe_code)]
+
+mod ssse3;
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+use std::is_x86_feature_detected;
+
+/// Arch-specific implementation of YCbCr conversion. Returns the number of pixels that were
+/// converted.
+pub fn color_convert_line_ycbcr(y: &[u8], cb: &[u8], cr: &[u8], output: &mut [u8]) -> usize {
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    #[allow(unsafe_code)]
+    {
+        if is_x86_feature_detected!("ssse3") {
+            unsafe {
+                return ssse3::color_convert_line_ycbcr(y, cb, cr, output);
+            }
+        }
+    }
+    return 0;
+}
+
+/// Arch-specific implementation of 8x8 IDCT.
+pub fn dequantize_and_idct_block_8x8(
+    coefficients: &[i16],
+    quantization_table: &[u16; 64],
+    output_linestride: usize,
+    output: &mut [u8],
+) {
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    #[allow(unsafe_code)]
+    {
+        if is_x86_feature_detected!("ssse3") {
+            unsafe {
+                ssse3::dequantize_and_idct_block_8x8(
+                    coefficients,
+                    quantization_table,
+                    output_linestride,
+                    output,
+                );
+                return;
+            }
+        }
+    }
+    unreachable!("No arch-specific IDCT available");
+}
+
+/// Returns true if an arch-specific IDCT is avaliable, false otherwise.
+pub fn has_arch_specific_idct() -> bool {
```

Comment:
In the spirit of 'parse, don't validate', I would return an optional, unsafe function pointer here instead of a bool. This places the detection code much closer to the specific requirements that the unsafe execution requires. It's easier to discern that we return (and execute) an `sss3` function only when `is_x86_feature_detected` is true.

Reply:
What would this do when there's no SIMD option? I guess return a None?

This could work, but I'm not sure if the extra indirection might cause some speed problems. Probably not, but I'll have to test.

Reply:
Returning a function point (`fn(…) -> …`)–or it should be more aptly named function reference as it must be non-null—is not double indirection. That's the exact type of the symbol of a function.

Reply:
Done, no slowdown I could measure :)